File: numbl_test_scripts/functions/js_user_functions/mynorm.js
// native: mynorm
// wasm: mynorm
register({
  check: function (argTypes, nargout) {
    return { outputTypes: [IType.num()] };
  },
  apply: function (args, nargout) {
    var x = args[0];
    var data = x.data; // Float64Array

    if (native) {
      // Native path: koffi can pass Float64Array directly as a pointer
      var mynorm_native = native.func("double mynorm(int n, double *x)");
      return RTV.num(mynorm_native(data.length, data));
    }

    if (wasm) {
      // Wasm path: must copy data into wasm linear memory since wasm
      // cannot directly access JS typed arrays
      var n = data.length;
      var ptr = wasm.exports.alloc_doubles(n);
      var mem = new Float64Array(wasm.exports.memory.buffer, ptr, n);
      mem.set(data);
      var result = wasm.exports.mynorm(n, ptr);
      wasm.exports.free_doubles();
      return RTV.num(result);
    }

    // JS fallback
    var sum = 0;
    for (var i = 0; i < data.length; i++) {
      sum += data[i] * data[i];
    }
    return RTV.num(Math.sqrt(sum));
  },
});

File: numbl_test_scripts/functions/js_user_functions/mynorm.c
/* L2 norm of a double array.
 *
 * For wasm:   emcc mynorm.c -O2 -s STANDALONE_WASM --no-entry -o mynorm.wasm
 * For native: gcc -shared -fPIC -O2 -o mynorm.so mynorm.c -lm
 */

#include <math.h>

#ifdef __EMSCRIPTEN__
#define EXPORT __attribute__((export_name("mynorm")))
#define EXPORT_ALLOC __attribute__((export_name("alloc_doubles")))
#define EXPORT_FREE __attribute__((export_name("free_doubles")))
#else
#define EXPORT
#define EXPORT_ALLOC
#define EXPORT_FREE
#endif

EXPORT
double mynorm(int n, const double *x) {
    double sum = 0.0;
    for (int i = 0; i < n; i++) {
        sum += x[i] * x[i];
    }
    return sqrt(sum);
}

/* Simple bump allocator for wasm — lets JS copy data into linear memory. */
static unsigned char heap[1 << 20]; /* 1 MB */
static int heap_offset = 0;

EXPORT_ALLOC
double *alloc_doubles(int n) {
    /* Align to 8 bytes */
    int aligned = (heap_offset + 7) & ~7;
    double *ptr = (double *)(heap + aligned);
    heap_offset = aligned + n * (int)sizeof(double);
    return ptr;
}

EXPORT_FREE
void free_doubles(void) {
    heap_offset = 0;
}
